### ISSUE TITLE
Feat: Sync playback between vlc instances on networked devices

### ISSUE THREAD

**Issue by ABelliqueux** [mrkeuz/vlcsync#2]
Hi, 

I've successfully set vlcsync up on a rpi and it works as intended.

I was wondering how trivial it would be to implement the possibility to synchronize playback between VLC instances
running on different computers on the same network.

e.g : server is at 192.168.0.10, computer 1 is at 192.168.0.11, computer 2 is at 192.168.0.12
each computer has the video files on their filesystem. 
`vlcsync` runs on server, `vlc` instances run on computer 1 and 2.  

E.G : On server, run `vlcsync`
         On computer 1, run `vlc --rc-host 192.168.0.10 SomeMedia1.mkv`
         On computer 2, run `vlc --rc-host 192.168.0.10 SomeMedia2.mkv`

One of the `vlc` instances could then be controlled via telnet, and synchronization would occur via `vlcsync`on the other instance.

**Comment by mrkeuz** [comment]
Hi, thanks for proposal, 

Yes, I thought about this, even I have [roadmap](https://github.com/mrkeuz/vlcsync/blob/main/DEVELOPMENT.md#roadmap) for this. It needs little refactoring for this. Currently, I store all controlled VLC by key (pid + port). It requires refactoring to host + port.

Plus I'm not finally decide how implement remote sync.

Problem 1: Closed random ports. If you start remote vlc's it started on random port. Usually firewall enabled and of course random ports blocked by default. So you require caring about open ports in firewall. Especially on windows. I don't want care about port numbers, it is not handy.

Problem 2: Discovery remote instances. There is a question about discovery remote VLC instances. Currently, I have access to process list and can view listen ports i.e. "Find all VLC processes and show all LISTEN ports". This is how I find all the players for sync. So in this case we should "ping" every port, check output, etc...

My current thought about this it is just run "another" vlcsync on remote as agent. So the agent will local discovery VLC's, then do synchronization "vlcsync ↔ vlcsync". In both direction when some vlc seek, etc.
 
Something like:
![vlcsync drawio](https://user-images.githubusercontent.com/6247921/180239665-72771ae0-a94f-43b3-ac70-85150155a7ea.png)

In this case: 
We require only one open port. And possible sync more two sources just add it to "network". 

Despite this, probably, I will implement both variants. Remote vlc + remote agent style sync. Maybe I will implement just static remote instance at first.

I need to evaluate all the options a bit and lay all the options in my head.
And some free time.

**Comment by ABelliqueux** [comment]
Hi, 
I missed that roadmap section, great to hear this is planned !

IMHO, your remote agent solution sounds more simple (one IP:PORT per remote host, and theorically allows you to add an infinite number of remote hosts, with multiple VLC instances each.

Maybe starting with a solution where you have to specify those variables (remote IP:port) manually could be less work at first ?

Some thoughts about this solution ; 
  * You could have an option overriding the port used, e.g: `vlcsync --port 9999` (meaning you use `vlc --telnet-port 9999` on the remote hosts. 
  * You could have a flag  to specify the remote hosts IPs or IP range,  manually, e.g : `vlcsync --remote-host 192.168.0.11 --remote-host 192.168.1.0/24`. This would prevent from having to scan each IP on the network.
  * You could have an option to choose to not look for vlc instances running on the same machine as vlcsync, e.g: `vlcsync --no-local-discover --remote-host ...`
  * You should be able to choose which network interface should be used, since you could have different subnets for each interface ( eth0 vs wlan0, etc. ) and not want to scan for instances on all of them...

Well these are just ideas, I'm not sure how trivial/possible it would be to do these kind of things, what's more having to make it work on windows too.

Thanks for your reply !

**Comment by mrkeuz** [comment]
Honestly, you inspire me to do some work 👍.

I've already does refactor. And almost done remote VLC support. Remains add command line parameters. Decide started from this feature, as it is easily to implement. Good start I think.

About your notes:

- Static remote, yes, even almost done. Will looks like i.e. for two remotes:
  ```shell
  vlcsync --rc-host <ip:port> --rc-host <ip:port>
  ```
- Disable local discover `--no-local-discover`, agree, yes, also was thought about it, will implement in this release. 

- About remote discovery, I will think about again in the future. Just require some protocol for sync events. During working, have decided just little postpone this feature as static covers all possible cases. Just need little more scripting. For starts remote VLC on certain port, open port if required (or tunnel it to local). So remote agent/discovery postpone currently and maybe done in the future. 

Static VLC's almost done, just need add command line args. And test it.

**Comment by ABelliqueux** [comment]
Wow, it's great that you found the time to do it ! Thank you for the effort you put into it.
I can beta test for you if you want, let me know if I can be of any help.

**Comment by mrkeuz** [comment]
Hi @ABelliqueux, I finally finish remote sync feature. You can test it if you want. Working brilliant! At least on first tests. 

The main thing I want to share. In process of work I figured out how to make remote discovery (without an agent on the remote machine) and even, what is more important without opening ports. Remote VLC's can continue run on random ports with algorithm:

- we connect via ssh to remote
- via i.e. `netstat -4nlp` get all VLC's ports, even VLC on `127.0.0.1` or `127.0.0.42` and not wide open
- then we forward port from remote port to local
- and connect to the local (forwarded) port and control VLC instance

Profit! 

Think should work. Even TIL windows has built-in OpenSSH server and can to forward ports (I haven’t worked with windows for a long time, I can’t give more precise details)

So in this case we not need distributed `vlcsync` etcetera. So don't need to code and support complex solution with distributed architecture.

Pros:
- All remote port closed 
- VLC is still can run on random port (even not wide "0.0.0.0")
- Even Windows has builtin ssh and this trick with port forwarding will work even in windows
- Simple solution that no need to effort for code and errors so fewer bugs

Thoughts?

**Comment by mrkeuz** [comment]
Fixed via #4 (`vlcsync v0.2.0`)
